File: CTCI/Ch2LinkedLists/Py/sum_lists.py
"""
Given two linked lists of numbers
sum them and print the result
as another linked list
"""


class Node(object):
    def __init__(self, next_node=None, data=None):
        self.next_node = next_node
        self.data = data


class LinkedList(object):
    def __init__(self, head=None):
        self.head = head

    @property
    def print_list(self):
        iterator = self.head

        while iterator is not None:
            print('{}-->'.format(iterator.data), end='')
            iterator = iterator.next_node

    def insert(self, data):
        new_node = Node(None, data)
        if self.head is None:
            self.head = new_node
            return
        new_node.next_node = self.head
        self.head = new_node


def sum_lists(l1, l2):
    list1 = []
    list2 = []
    temp1 = l1.head
    temp2 = l2.head
    while temp1 is not None:
        list1.append(temp1.data)
        temp1 = temp1.next_node
    while temp2 is not None:
        list2.append(temp2.data)
        temp2 = temp2.next_node

    fl1 = ''.join(str(x) for x in list1)
    fl2 = ''.join(str(x) for x in list2)

    total = int(fl1) + int(fl2)
    nl = LinkedList(None)

    for val in str(total):
        nl.insert(val)

    nl.print_list


def main():
    l1 = LinkedList(None)
    l2 = LinkedList(None)
    vals1 = [7, 1, 6]
    vals2 = [5, 9, 2]

    for v in vals1:
        l1.insert(v)

    for v in vals2:
        l2.insert(v)

    l1.print_list
    print('\n')
    l2.print_list

    print('\n')
    sum_lists(l1, l2)


main()

File: CTCI/Ch2LinkedLists/java/SumLists.java
import java.util.*;
import java.lang.*;
/**
  * This program takes two linked lists and 
  * sums their total values in reverse and normal
  * order.
  * Example:
  * 6-->1-->7 + 7-->1-->9 = 1336 (reversed)
  * 7-->1-->6 + 9-->1-->7 = 1336 (non-reversed)
  */

public class SumLists {

	protected Node head;

	public class Node {
		int data;
		Node next;

		public Node() {
			next = null;
		}

		public Node(int nodeData) {
			data = nodeData;
		}
	}

	public void insert(int data) {
		if (head == null) {
			Node newNode = new Node(data);
			head = newNode;
		} else {
			Node newNode = new Node(data);
			newNode.next = head;
			head = newNode;
		}
	}

	public void print(Node head) {
		Node temp = head;
		while (temp != null) {
			System.out.print(temp.data + "-->");
			temp = temp.next;
		}
		System.out.println("\n");
	}
	
	public int sumLists(Node root1, Node root2, boolean reverse) {
		StringBuilder final1 = new StringBuilder();
		StringBuilder final2 = new StringBuilder();
		Node temp1 = root1;
		Node temp2 = root2;

		while(temp1 != null) {
			final1.append(temp1.data);
			temp1 = temp1.next;
		}

		while(temp2 != null) {
			final2.append(temp2.data);
			temp2 = temp2.next;
		}

		if (reverse) {
			final1 = final1.reverse();
			final2 = final2.reverse();
			return Integer.parseInt(final1.toString()) + Integer.parseInt(final2.toString());
		}
		else {
			return Integer.parseInt(final1.toString()) + Integer.parseInt(final2.toString());
		}
	}

	public static void main(String[] args) {
		SumLists s1 = new SumLists();
		s1.insert(6);
		s1.insert(1);
		s1.insert(7);
		SumLists s2 = new SumLists();
		s2.insert(7);
		s2.insert(1);
		s2.insert(9);
		SumLists s3 = new SumLists();
		s3.insert(7);
		s3.insert(1);
		s3.insert(6);
		SumLists s4 = new SumLists();
		s4.insert(9);
		s4.insert(1);
		s4.insert(7);
		SumLists s = new SumLists();
		System.out.println(s.sumLists(s1.head, s2.head, true));
		System.out.println(s.sumLists(s3.head, s4.head, false));
	}
}
